src/osd/PrimaryLogPG.cc

```diff
@@ -8501,16 +8501,26 @@ void PrimaryLogPG::_do_rollback_to(OpContext *ctx, ObjectContextRef rollback_to,
   t->clone(soid, rollback_to_sobject);
   t->add_obc(rollback_to);
 
+  if (obs.oi.size > 0) {
+    // change most recent clone_overlap to [0, obs.oi.size] before rollback to snap
+    if (ctx->new_snapset.clones.size() > 0) {
+      interval_set<uint64_t> rollback_overlaps;
+      rollback_overlaps.insert(0, obs.oi.size);
+      interval_set<uint64_t> &newest_overlap =
+        ctx->new_snapset.clone_overlap.rbegin()->second;
+      newest_overlap.swap(rollback_overlaps);
+    }
+
   map<snapid_t, interval_set<uint64_t> >::iterator iter =
     snapset.clone_overlap.lower_bound(snapid);
   ceph_assert(iter != snapset.clone_overlap.end());
   interval_set<uint64_t> overlaps = iter->second;
   for ( ;
 	iter != snapset.clone_overlap.end();
-	++iter)
-    overlaps.intersection_of(iter->second);
+	++iter) {
+        overlaps.intersection_of(iter->second);
+      }
 
-  if (obs.oi.size > 0) {
     interval_set<uint64_t> modified;
     modified.insert(0, obs.oi.size);
     overlaps.intersection_of(modified);
```

Comment:
This looks right to me.
- overlaps is set of byte ranges that didn't change between the rollback target and head.
- intersection(overlaps, [0, obs.oi.size)) is the subset of the above which falls within the size of the current object
- [0, obs.oi.size) - intersection(overlaps, [0, obs.oi.size)) is therefore the set of ranges modified between the rollback target and head
- ctx->modified_ranges.union_of(modified); is the above plus any other extents modified in this transaction.

If you are seeing a bug with clone_overlaps, presumably one of the above steps is incorrect.

Reply:
yes, I found the latest overlap is not that correct when rollback_to snap.

```shell
[root@test xingd]# rbd create ceph-vm-pool-1/test_rollback -s 4M
[root@test xingd]# fio -ioengine=rbd -bs=4M -size=4M -rw=write -pool=ceph-vm-pool-1 -rbdname=test_rollback --group_report -name=test -direct=1 -numjobs=1 -iodepth=1
[root@test xingd]# rbd snap create ceph-vm-pool-1/test_rollback@base
[root@test xingd]# fio -ioengine=rbd -bs=2M -size=2M -rw=write -pool=ceph-vm-pool-1 -rbdname=test_rollback --group_report -name=test -direct=1 -numjobs=1 -iodepth=1
[root@test xingd]# rados -p ceph-vm-pool-1 listsnaps rbd_data.b4bf851b702e15.0000000000000000
rbd_data.b4bf851b702e15.0000000000000000:
cloneid	snaps	size	overlap
48	48	4194304	[2097152~2097152]
head	-	4194304
[root@test xingd]# rbd snap rollback ceph-vm-pool-1/test_rollback@base
[root@test xingd]# rados -p ceph-vm-pool-1 listsnaps rbd_data.b4bf851b702e15.0000000000000000
rbd_data.b4bf851b702e15.0000000000000000:
cloneid	snaps	size	overlap
48	48	4194304	[2097152~2097152]           ### after rollback， the overlap shoud be [0~4194304]
head	-	4194304
```

another test is:
```shell
[root@test xingd]# rbd create ceph-vm-pool-1/test_rollback1 -s 4M
[root@test xingd]# fio -ioengine=rbd -bs=4M -size=4M -rw=write -pool=ceph-vm-pool-1 -rbdname=test_rollback1 --group_report -name=test -direct=1 -numjobs=1 -iodepth=1
[root@test xingd]# rbd snap create ceph-vm-pool-1/test_rollback1@base
[root@test xingd]# rados -p ceph-vm-pool-1 listsnaps rbd_data.a9d86dcb31684.0000000000000000
rbd_data.a9d86dcb31684.0000000000000000:
cloneid	snaps	size	overlap
head	-	4194304
[root@test xingd]# rbd snap rollback ceph-vm-pool-1/test_rollback1@base
[root@test xingd]# rados -p ceph-vm-pool-1 listsnaps rbd_data.a9d86dcb31684.0000000000000000
rbd_data.a9d86dcb31684.0000000000000000:
cloneid	snaps	size	overlap
103	103	4194304	[0~4194304]                ### write nothing after snap create, and rollback to snap, the latest overlap is [0～4194304] （which is correct）
head	-	4194304
```

Reply:
Ok, the first step would be to create a rados unit test in test/librados/snapshots_cxx.cc which demonstrates the problem.  Be sure to include cases where:
1. the rollback target is the newest clone
2. the rollback target is the most recent clone
3. the rollback target is the oldest clone
4. the rollback target is neither the oldest nor the newest clone

I think the actual bug is that the above loop should skip the last overlap entry.  The rollback overwrites the head, so we don't care about the overlap between head and the most recent snapshot -- we want the overlap between the most recent *clone* and the rollback target.  Once you have a unit test, try changing that.

Reply:
> This looks right to me.
> 
>     * overlaps is set of byte ranges that didn't change between the rollback target and head.

In other words, this part is wrong.  We want overlap to be the set of byte ranges that didn't change between the rollback target and the *newest clone*, not head.